src/jit/gentree.h

```diff
@@ -2199,6 +2198,92 @@ struct GenTree
     inline GenTree(genTreeOps oper, var_types type DEBUGARG(bool largeNode = false));
 };
 
+struct GenTreePhi : public GenTree
+{
+    struct Use
+    {
+        GenTree* op;
+        Use*     next;
+
+        Use(GenTree* op, Use* next = nullptr) : op(op), next(next)
+        {
+        }
+    };
+
+    class UseIterator
+    {
+        Use* m_use;
+
+    public:
+        UseIterator(Use* use) : m_use(use)
+        {
+        }
+
+        Use& operator*() const
+        {
+            return *m_use;
+        }
+
+        Use* operator->() const
+        {
+            return m_use;
+        }
+
+        UseIterator& operator++()
+        {
+            m_use = m_use->next;
+            return *this;
+        }
+
+        bool operator==(const UseIterator& i) const
+        {
+            return m_use == i.m_use;
+        }
+
+        bool operator!=(const UseIterator& i) const
+        {
+            return m_use != i.m_use;
+        }
+    };
+
+    class UseList
+    {
+        Use* m_uses;
+
+    public:
+        UseList(Use* uses) : m_uses(uses)
+        {
+        }
+
+        UseIterator begin() const
+        {
+            return UseIterator(m_uses);
+        }
+
+        UseIterator end() const
+        {
+            return UseIterator(nullptr);
+        }
+    };
+
+    Use* gtUses;
+
+    GenTreePhi(var_types type) : GenTree(GT_PHI, type), gtUses(nullptr)
+    {
+    }
+
+    UseList Uses()
+    {
+        return UseList(gtUses);
+    }
+
+#if DEBUGGABLE_GENTREE
```

Comment:
What is the purpose of `DEBUGGABLE_GENTREE`?

Reply:
Every `GenTree` class needs a default constructor to keep `GenTree::VtablePtr GenTree::GetVtableForOper(genTreeOps oper)` happy. It declares variables of various `GenTree` types so it can get its hands on a vtable pointer to use when the class of a `GenTree` node is changed using to `ChangeOper`. Of course, it's all a rather crazy, non-standard, not-anything hack :smile:

Reply:
From a user perspective, it allows the VS debugger to show you the "real" subtype of a `GenTree*` variable, which is super useful. Maybe something that is done in a more standard way with RTTI?

Reply:
I don't know, the standard way would be to use placement `new` to change the class of an object. But... I need to double check the standard... I think doing that will invalidate the existing object's contents and we don't want that when using `ChangeOper`.

Considering that this is debug only code I think it's fine as it is. Perhaps we should just add a macro to declare these default constructors on a single line, just to reduce debug noise.

Reply:
One "problem" with `DEBUGGABLE_GENTREE` is it isn't available in Release, so not in production crash dumps, but making it available there was considered too expensive, and unnecessary.

Reply:
> but making it available there was considered too expensive, and unnecessary

Yeah, at the end of day, `DEBUGGABLE_GENTREE` is pure convenience. One can still figure out the class from `gtOper`, it's just more work to do so in the debugger. I suppose it would be cool if C++'s RTTI was somehow customizable but it isn't, it requires a vtable pointer to work.

Reply:
Maybe consider adding an operand count? There may be places we'd prefer to avoid traversing large PHI lists.